solidity/contracts/BancorNetwork.sol

```diff
@@ -218,79 +215,62 @@ contract BancorNetwork is IBancorNetwork, TokenHolder, ContractRegistryClient, F
+        // iterate over the conversion data
+        for (uint256 i = 0; i < _data.length; i++) {
+            ConversionStep memory stepData = _data[i];
 
-            // if the smart token isn't the source (from token), the converter doesn't have control over it and thus we need to approve the request
-            if (_path[i - 1] != _path[i - 2] && _path[i - 2] != IERC20Token(0))
-                ensureAllowance(_path[i - 2], converter, fromAmount);
+            // newer converter
+            if (stepData.isV27OrHigherConverter) {
+                // transfer the tokens to the converter only if the network contract currently holds the tokens
+                // not needed with ETH or if it's the first conversion step
+                if (i != 0 && _data[i - 1].beneficiary == address(this) && !etherTokens[stepData.sourceToken])
+                    safeTransfer(stepData.sourceToken, stepData.converter, fromAmount);
+            }
+            // older converter
+            // if the source token is the smart token, no need to do any transfers as the converter controls it
+            else if (stepData.sourceToken != stepData.smartToken) {
+                // grant allowance for it to transfer the tokens from the network contract
+                ensureAllowance(stepData.sourceToken, stepData.converter, fromAmount);
+            }
 
-            // make the conversion - if it's the last one, also provide the minimum return value
-            toAmount = change(converter, _path[i - 2], _path[i], fromAmount, i == lastIndex ? _minReturn : 1);
+            // do the conversion
+            if (!stepData.isV27OrHigherConverter)
+                toAmount = ILegacyBancorConverter(stepData.converter).change(stepData.sourceToken, stepData.targetToken, fromAmount, stepData.minReturn);
+            else if (etherTokens[stepData.sourceToken])
+                toAmount = stepData.converter.convertInternal.value(msg.value)(stepData.sourceToken, stepData.targetToken, fromAmount, stepData.minReturn, stepData.beneficiary);
+            else
+                toAmount = stepData.converter.convertInternal(stepData.sourceToken, stepData.targetToken, fromAmount, stepData.minReturn, stepData.beneficiary);
 
             // pay affiliate-fee if needed
-            if (address(_path[i]) == bntToken) {
+            if (stepData.processAffiliateFee) {
                 uint256 affiliateAmount = toAmount.mul(_affiliateFee).div(AFFILIATE_FEE_RESOLUTION);
-                require(_path[i].transfer(_affiliateAccount, affiliateAmount));
+                require(stepData.targetToken.transfer(_affiliateAccount, affiliateAmount));
                 toAmount -= affiliateAmount;
-                bntToken = address(~0);
             }
 
-            emit Conversion(_path[i - 1], _path[i - 2], _path[i], fromAmount, toAmount, msg.sender);
+            emit Conversion(stepData.smartToken, stepData.sourceToken, stepData.targetToken, fromAmount, toAmount, msg.sender);

```

Comment:
I know that you've heard me saying this before, but consider removing this event because it contains redundant information (already provided in the `BancorConverter` layer), which is also partial information (does not include the fee).
Another incentive for removing this event, is the fact that it is emitted inside the loop, which can actually save quite a bit of gas.

Reply:
I want to redesign the events to make it REALLY easier to get on-chain data since I do agree we did half a solution here, so let's keep it until we have a better alternative - this is right now helping some of the data aggregators.
Maybe we can even do it in this version, though it's getting tight :)